### PR TITLE
Ordered Transform

### PR DESCRIPTION
**Pull request by bwengals**
Transform to constrain a random vector's elements to be nondecreasing.  First commit thanks to @aseyboldt.  I'd also like to include an equivalent to Stan's `ordered_logistic` distribution in this PR.

### CODE REVIEW

File: pymc3/distributions/transforms.py
Comment (junpenglao) [inline pymc3/distributions/transforms.py]: `draw_values` was already imported from `distribution` earlier in the file.

**Comment by fonnesbeck** [comment]
Maybe add WIP to the title if you are still looking to add code.

**Comment by bwengals** [comment]
What do you guys think of an API similar to `Bound`? 

    OrderedPrior = pm.Ordered(pm.Flat, 'increasing')

Maybe this makes it more straightforward to compose transformations, and the user doesn't need to specify a `transform` arg on a distribution.

**Comment by junpenglao** [comment]
Do you mean something like the `Bound`? which you do:
```python
OrderedPrior = pm.Ordered(pm.Gamma, 'increasing')
with pm.Model():
    ...
    a = OrderedPrior('a', alpha=1., beta=1.)
```
In that case, instead of a composite transform you do this?:
```python
OrderedPrior = pm.Ordered(pm.Gamma, 'increasing')
with pm.Model():
    ...
    a = OrderedPrior('a', alpha=1., beta=1., 
                                 transform=pm.distributions.transforms.LogExpM1())
```

**Comment by bwengals** [comment]
Yes, that's what I meant.  Although it's implemented as a transform, it's an additional constraint on a prior in the same way `Bound` is.  Then you can handle composing multiple transforms within the `Ordered` object and the composite transform isn't directly exposed to users.

**Comment by junpenglao** [comment]
I like your idea. Would it be better to put it in the same file as `Bound`?

**Comment by twiecki** [comment]
Ping @bwengals.

**Comment by junpenglao** [comment]
FYI: the current version does not work for 2d array. I did some edit here: https://discourse.pymc.io/t/order-statistics-in-pymc3/617/3?u=junpenglao
Not sure if we can make it work for both 1d and 2d tho...

**Comment by bwengals** [comment]
Whoops just saw the ping.  I started looking at this again the last couple days, been working on the OrderedLogistic distribution part.  

I'm been having a hard time getting the ordered transform to work like `Bound` though.  Would it be alright to leave them in just as transforms?

Not sure about the 1D vs 2D transform thing, will look at that

**Comment by junpenglao** [comment]
> I'm been having a hard time getting the ordered transform to work like Bound though. Would it be alright to leave them in just as transforms?

Agree. In that case could you please also add `Composed` transforms https://discourse.pymc.io/t/mixture-models-and-breaking-class-symmetry/208/5?u=junpenglao?

**Comment by bwengals** [comment]
Ok awesome :), yup, `Composed` is in here.

**Comment by junpenglao** [comment]
Seems there would be an error in the Jacobian determinant when we are using `Composite` with `Transform` (see discussion here: https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/7?u=junpenglao)
cc @aseyboldt

**Comment by bwengals** [comment]
sorry for slowness, been having some difficulty getting the tests to work on the ordered/composite transforms.  Just want to status update anyone watching this

**Comment by bwengals** [comment]
I've been stuck on getting these tests to pass :(.  Anyone have any experience with the transforms testing framework?

File: pymc3/distributions/transforms.py
Comment (junpenglao) [inline pymc3/distributions/transforms.py]: Are you sure this is not forward?? See
https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/12?u=junpenglao

File: pymc3/tests/test_transforms.py
Comment (junpenglao) [inline pymc3/tests/test_transforms.py]: Something like:
```python
def test_ordered_vals():
    vals = get_values(tr.ordered, Vector(R, 3),
                      tt.dvector, np.zeros(3))
    close_to_logical(np.diff(vals) >= 0, True, tol)
```
Again, this test shows that the forward and backward is reversed in your current implementation.

File: pymc3/tests/test_transforms.py
Comment (junpenglao) [inline pymc3/tests/test_transforms.py]:
```python
def test_ordered_jacobian_det():
    check_jacobian_det(tr.ordered, Vector(R, 2),
                      tt.dvector, np.array([0, 0]), elemwise=False)
```

**Review by junpenglao — CHANGES_REQUESTED** [review]

**Comment by junpenglao** [comment]
Also need tests for the `Composed` transformation

File: pymc3/tests/test_transforms.py
Comment (junpenglao) [inline pymc3/tests/test_transforms.py]: The domain of ordered transform is not real value, but sorted real value. You should create a new domain `SortedVector` here.

Comment (junpenglao) [reply pymc3/tests/test_transforms.py]: Maybe something like:
```python
# maybe this could be put into test_distributions.py
def SortedVector(n):
    vals = []
    np.random.seed(42)
    for _ in range(10):
        vals.append(np.sort(np.random.randn(n)))
    return Domain(vals, edges=(None, None))

def test_ordered():
    check_vector_transform_identity(tr.ordered, SortedVector(6))
```

**Comment by bwengals** [comment]
ah thanks a ton @junpenglao, that completely cleared things up for me!

**Comment by fonnesbeck** [comment]
Definitely add this to the change log before merging.

**Comment by twiecki** [comment]
@bwengals Is this finished? Would be great to merge (needs rebase and release-notes).

**Comment by junpenglao** [comment]
Need to add tests for the `Composed` transformation and release note.

File: pymc3/tests/test_distributions.py
Comment (ColCarroll) [inline pymc3/tests/test_distributions.py:345]: `sys.float_info.epsilon`: neat!

**Comment by junpenglao** [comment]
@bwengals any update?

**Comment by bwengals** [comment]
last I picked it up, got stuck on getting the jacobian tests for the `Composed` transform working.  Everything else is good to go, included the `OrderedLogistic` likelihood.  Been seeing what you guys have in this https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/14

I suppose we could have `Composed` be a separate PR, and open an issue for it in the meantime.  I can take another crack at it this weekend too and see.  What do you think?

**Comment by junpenglao** [comment]
No problem! Also you can have a look at the composite transformation in tensorflow: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/bijectors/chain.py

**Review by junpenglao — APPROVED** [review]

**Comment by junpenglao** [comment]
@bwengals hope you dont mind I go ahead and fixed the test (need to use the ordered transformation soonish myself ;-) )

**Comment by junpenglao** [comment]
Please do not merge yet - I think there is some potential problems of the Jacobian.

**Comment by bwengals** [comment]
Thanks for bailing me out!

**Comment by junpenglao** [comment]
OK this is now ready for review.

**Comment by twiecki** [comment]
Also needs a rebase.

**Comment by junpenglao** [comment]
Can I even do that? This is not my branch.
[EDIT] Done.

File: pymc3/distributions/transforms.py
Comment (twiecki) [inline pymc3/distributions/transforms.py:250]: This looks trivial to have figured out :P.

Comment (junpenglao) [reply pymc3/distributions/transforms.py:250]: I know - it blows my mind when I came across this.

**Comment by junpenglao** [comment]
I did some additional tests and it should be fine for this PR now.

I will submit another PR for the chain transformation and add some additional tests.

**Comment by twiecki** [comment]
Congrats @junpenglao and @bwengals, quite a tricky one!

**Pull request merged**
